ksp 対応

dYoshitakaKato opened the pull request:
## 目的
`fieldname-constant-annotation:compiler` を ksp に対応する

## やったこと
- `fieldname-constant-annotation:compiler` で `kotlin` をサポートするように変更
- `ksp` を導入した
- `@DefineNames` を `DefineNamesProcessor.java` と同様に処理する `ksp` 対応の `DefineNamesProvider` を作成した
- `DefineNamesProcessorTest.java` と同等のテストをする `DefineNamesProviderTest` を作成した
- 古い `DefineNamesProcessor`, `DefineNamesProcessorTest`, `javax.annotation.processing.Processor` を削除した

## 動作確認
- ユニットテストが全て通ることを確認
- `publishToMavenLocal` 後に導入プロジェクトで下記変更後、ビルドに成功することを確認
	- `kapt("com.lisb.fieldnameconstant-annotation:compiler:2.0.0-SNAPSHOT")` を `ksp("com.lisb.fieldnameconstant-annotation:compiler:2.0.0-SNAPSHOT")` へ変更

## 影響範囲
- `@DefineNames` があるクラスの `_C*` ファイル生成
- ユニットテスト

dYoshitakaKato committed:
DefineNamesProvider の作成

ksp で `@DefineNames` を処理する `DefineNamesProvider` を作成した。

dYoshitakaKato committed:
static フィールドの出力を選べるようにする

既存の `DefineNamesProcessor` では `static` フィールドの出力をプロパティによって、切り替えていた。
同様の動作となるように `DefineNamesProvider` にも `static` フィールドの出力を切り替える機能を追加した。

`Kotlin` では `static` フィールドは存在しないため `KSClassDeclaration.declarations` から `filterIsInstance<KSPropertyDeclaration>` を行い `static` を含めるようにした

dYoshitakaKato committed:
project に Kotlin サポートを追加

ksp を導入するため、 Kotlin を追加

dYoshitakaKato committed:
DefineNamesProviderTest を作成

既存の `DefineNamesProcessorTest` と同様のテストを行う `DefineNamesProviderTest` を作成した。
ksp 単体では UnitTest でファイルの生成まで行うことができなかったため `https://github.com/zacsweers/kotlin-compile-testing` を導入しファイルの生成までテストを行うようにした。
`https://github.com/tschuchortdev/kotlin-compile-testing` も検討したが ksp 2 が未サポートだったため採用しなかった。

dYoshitakaKato committed:
kapt での生成を削除

`ksp` での生成に対応したため、不要となった kapt の生成を削除

dYoshitakaKato committed:
ksp を導入

`com.google.devtools.ksp` を導入した

y-gomi112358 reviewed (approved):
ご対応ありがとうございます！問題ないように思います！

takara727 committed:
同じようなdependenciesやrepositoriesが複数のファイルに記載されていたのでsettings.gradleにまとめた

takara727 reviewed (commented):
同じような `dependencies` や `repositories` が複数のファイルに記載されていたので`settings.gradle` にまとめました。
別の branch で push して「こうしたほうが良いのでは」と記載するつもりが間違ってそのままうの branch に push してしまいました。申し訳ありません。

https://github.com/lisb/fieldname-constant-annotation/pull/4/commits/940ebe08f828231c9c06b0945f2f859893079496

takara727 commented on compiler/src/main/java/com/lisb/defname/internal/DefineNamesProvider.kt:
```diff
@@ -0,0 +1,60 @@
+                    val fieldName = annotation?.value ?: property.simpleName.asString()
+                    classWriter.addFields(fieldName)
+                }
+                val dependencies = Dependencies(false, requireNotNull(ksAnnotated.containingFile))
```
`sources` に先祖のクラスのファイルとかも指定しないと、先祖のクラスに変更があっても再ビルドされない気がするのですがいかがでしょうか?

dYoshitakaKato committed:
Fix:static出力対象時に親のプロパティが出力されない

`withStaticField = true` 時に親のプロパティが出力対象となっていなかったため修正
また、親の `static` プロパティを再起的に確認し出力するように変更した

dYoshitakaKato committed:
Fix:全ての親を依存関係にする

`Dependencies` に指定していたいのが `@DefineNames` があるファイルのみとなっていて、親ファイルが変更された場合にファイルが変更されない動作となっていた。
そのため、親も依存関係に含めるよう修正

dYoshitakaKato committed:
不要な gradlePluginPortal を削除

`gradlePluginPortal` は不要な参照なため削除

dYoshitakaKato commented on settings.gradle:
```diff
@@ -1,2 +1,24 @@
+        mavenCentral()
+        jcenter()
+        maven { url "https://maven.seasar.org/maven2" }
+        gradlePluginPortal()
```
ありがとうございます。ご指摘の通り不要でしたため、削除しました。
[`0c5a769` (#4)](https://github.com/lisb/fieldname-constant-annotation/pull/4/commits/0c5a769d8561846fd09f5f0c8f635d568a491fdc)

takara727 commented on compiler/src/main/java/com/lisb/defname/internal/DefineNamesProvider.kt:
```diff
@@ -0,0 +1,65 @@
+                val allFiles =
+                    ksAnnotated.containingFile?.let { parentFiles.plus(it) } ?: parentFiles
+                val dependencies =
+                    Dependencies(aggregating = parentFiles.any(), sources = allFiles.toTypedArray())
```
https://kotlinlang.org/docs/ksp-incremental.html#aggregating-vs-isolating を読む限り、 ここでは常に  `aggregating = false` じゃないかと思うのですがいかがでしょうか?

dYoshitakaKato replied:
ありがとうございます。確認しました `aggregating = false` が適切でした。修正し force push しました。
https://github.com/lisb/fieldname-constant-annotation/compare/0c5a769d8561846fd09f5f0c8f635d568a491fdc..a7e4e7160e2c6d860a2e4ce25b7e483a07e584ff

dYoshitakaKato commented on compiler/src/main/java/com/lisb/google/devtools/ksp/symbol/KSClassDeclarationExt.kt:
```diff
@@ -0,0 +1,14 @@
+object KSClassDeclarationExt {
+    fun KSClassDeclaration.getFields(withStaticField: Boolean): Sequence<KSPropertyDeclaration> {
+        return if (withStaticField) {
+            this.declarations.filterIsInstance<KSPropertyDeclaration>()
```
ありがとうございます。
https://github.com/lisb/fieldname-constant-annotation/compare/0c5a769d8561846fd09f5f0c8f635d568a491fdc..a7e4e7160e2c6d860a2e4ce25b7e483a07e584ff `static field` を再起的に取得するよう修正しました。またテストを追加し force push しました。

takara727 commented on compiler/src/main/java/com/lisb/google/devtools/ksp/symbol/KSClassDeclarationExt.kt:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    fun KSClassDeclaration.getAllFieldsIncludingStatic(): List<KSPropertyDeclaration> {
+        val fields = mutableListOf<KSPropertyDeclaration>()
+        val declarations = this.declarations.filterIsInstance<KSPropertyDeclaration>()
+        for (declaration in declarations) {
+            if (fields.any { it.simpleName.asString() == declaration.simpleName.asString() }) continue
+            fields.add(declaration)
+        }
+        for (parent in this.superTypes.mapNotNull { it.resolve().declaration as? KSClassDeclaration }) {
+            fields.addAll(parent.getAllParentFieldsIncludingStatic(fields))
+        }
+        return fields
+    }
+
+    private fun KSClassDeclaration.getAllParentFieldsIncludingStatic(
+        knownFields: List<KSPropertyDeclaration>
+    ): List<KSPropertyDeclaration> {
+        val fields = mutableListOf<KSPropertyDeclaration>()
+        val declarations = this.declarations.filterIsInstance<KSPropertyDeclaration>()
+        for (declaration in declarations) {
+            if (knownFields.any { it.simpleName.asString() == declaration.simpleName.asString() }) continue
+            if (declaration.modifiers.contains(Modifier.PRIVATE)) continue
+            fields.add(declaration)
+        }
+        for (parent in this.superTypes.mapNotNull { it.resolve().declaration as? KSClassDeclaration }) {
+            fields.addAll(parent.getAllParentFieldsIncludingStatic(knownFields + fields))
+        }
+        return fields
+    }
```
同じような内容が重複しています。
以下のようにしたほうがスッキリすると思うのですがいかがでしょうか?

```Kotlin
fun KSClassDeclaration.getAllFieldsIncludingStatic(): List<KSPropertyDeclaration> {
    return mutableListOf<KSPropertyDeclaration>().apply {
        getAllFieldsIncludingStatic(this, false)
    }
}

/**
 * @param isInherited 継承されたクラスか否か(継承されたクラスの場合、Private フィールドは無視する)
 */
private fun KSClassDeclaration.getAllFieldsIncludingStatic(
    fields: MutableList<KSPropertyDeclaration>,
    isInherited: Boolean
) {
    val declarations = this.declarations.filterIsInstance<KSPropertyDeclaration>()
    for (declaration in declarations) {
        if (fields.any { it.simpleName.asString() == declaration.simpleName.asString() }) continue
        if (isInherited && declaration.modifiers.contains(Modifier.PRIVATE)) continue
        fields.add(declaration)
    }
    val parent =
        this.superTypes.firstNotNullOfOrNull { it.resolve().declaration as? KSClassDeclaration }
    parent?.getAllFieldsIncludingStatic(fields, true)
}
```

dYoshitakaKato committed:
TestSource の名称を変更

別パッケージとなっている `com.lisb.defname.internal.TestSource1` と連番となっていた。
そのため、 `com.lisb.google.devtools.ksp.symbol` 内で連番となるように再番を修正

dYoshitakaKato committed:
kspProcess を作成し KSClassDeclarationExtTest で使用する ksp の処理を共通化

`testGetAllParentFiles`, `testGetFieldsWithStaticField`, `testGetFieldsWithOutStaticField` で使用する共通処理を `kspProcess` として抜き出し共通化した

dYoshitakaKato commented on compiler/src/test/java/com/lisb/google/devtools/ksp/symbol/TestSource1.kt:
ありがとうございます。
ご指摘の通り別パッケージで連番となるのはおかしいですね。`TestSource1`, `TestSource2` と新たに再番するように修正しました。
[`acb6064` (#4)](https://github.com/lisb/fieldname-constant-annotation/pull/4/commits/acb6064634ae9b62de865f88e23c384cc41e3431)

takara727 reviewed (approved):
問題ないと思います!
